### PR TITLE
feat: add `snow run` command for project scripts

### PR DESCRIPTION
**Pull request by sfc-gh-mborins**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.
   * [ ] I've described my changes in the documentation.

### Changes description
Add a new `snow run` command that allows users to define and execute project-specific scripts in snowflake.yml or manifest.yml. Features include:

- Load script from either snowflake.yml or manifest.yml with conflict detection
- Track script source file and show in --list output
- Script definitions with cmd, description, shell, cwd, and env options
- Composite scripts using `run:` to sequence multiple scripts
- Variable interpolation with ${env.VAR} syntax
- CLI overrides with -D flag
- Dry-run mode and continue-on-error for composite scripts
- Extensive unit and integration tests covering various scenarios and edge cases

### CODE REVIEW

File: src/snowflake/cli/_plugins/run/commands.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    """
+    ctx = get_cli_context()
+
+    project_root = ctx.project_root
+    if project_root is None:
+        project_root = Path.cwd()
+
+    manager = ScriptManager(project_root)
+
+    if list_scripts:
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/_plugins/run/commands.py:104]: Will this work with `--format=json`?

File: src/snowflake/cli/_plugins/run/commands.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                cc.message(f"  {name}")
+            return MessageResult("\nSpecify a script name to run, or use --list")
+        return MessageResult(
+            "No scripts defined. Add a 'scripts' section to snowflake.yml or manifest.yml"
+        )
+
+    vars_dict = {}
+    if var_overrides:
+        for v in parse_key_value_variables(var_overrides):
+            vars_dict[v.key] = v.value
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/_plugins/run/commands.py]:
```suggestion
vars_dict = {v.key: v.value for v in parse_key_value_variables(var_overrides)}
```
More pythonic

File: src/snowflake/cli/_plugins/run/commands.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    script = manager.get_script(script_name)
+    if not script:
+        available = list(manager.list_scripts().keys())
+        if available:
+            raise ClickException(
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/commands.py]: Please raise `CliError`

File: src/snowflake/cli/_plugins/run/manager.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        manifest_scripts, manifest_source = self._load_manifest_scripts()
+
+        if snowflake_scripts and manifest_scripts:
+            raise ClickException(
+                "Scripts defined in both manifest.yml and snowflake.yml.\n"
+                "Scripts must be defined in only one file per directory.\n\n"
+                "Recommendation: Move all scripts to one file.\n"
+                "- Use manifest.yml for DCM-focused projects\n"
+                "- Use snowflake.yml for app-focused projects"
+            )
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/manager.py]: Thinking out loud - is that a bad thing really? Maybe we can merge scripts from both sources?

File: src/snowflake/cli/_plugins/run/commands.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        scripts = manager.list_scripts()
+        if not scripts:
+            return MessageResult("No scripts defined in snowflake.yml or manifest.yml")
+
+        source = manager.scripts_source or "project"
+        cc.message(f"Available scripts (from {source}):")
+        for name, script in scripts.items():
+            desc = script.description or "(no description)"
+            cc.message(f"  {name:20} {desc}")
+        return MessageResult("")
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/commands.py]: You can also use EmptyResult

File: src/snowflake/cli/_plugins/run/commands.py (hunk 1 above)
Comment (sfc-gh-jwilkowski) [reply src/snowflake/cli/_plugins/run/commands.py:104]: Would be better to use CollectionResult/ObjectResult to have support for different formatting options. I don't think json/csv would be very useful in this case, but having some response in those formats is better than nothing

Comment (sfc-gh-jwilkowski) [reply src/snowflake/cli/_plugins/run/commands.py:104]: Another issue I see is that introduces an inconsistency with other cli commands. As raised before, we usually have a dedicated command for listing objects

Comment (sfc-gh-jwilkowski) [reply src/snowflake/cli/_plugins/run/commands.py:104]: with a simple script like 
```
scripts:
  show_pwd:
    cmd: "pwd"
    description: "run pwd"
```
we get this:
```
snow run show_pwd --format json
/Users/jwilkowski/snowflake/dcm_test/dcm_manifest_v2
{
    "message": ""
}
```

File: src/snowflake/cli/_plugins/run/commands.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    if list_scripts:
+        scripts = manager.list_scripts()
+        if not scripts:
+            return MessageResult("No scripts defined in snowflake.yml or manifest.yml")
+
+        source = manager.scripts_source or "project"
+        cc.message(f"Available scripts (from {source}):")
+        for name, script in scripts.items():
+            desc = script.description or "(no description)"
+            cc.message(f"  {name:20} {desc}")
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/commands.py]: Use `sanitize_for_terminal()` before printing anything that comes from un untrusted source

File: src/snowflake/cli/_plugins/run/commands.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        scripts = manager.list_scripts()
+        if scripts:
+            source = manager.scripts_source or "project"
+            cc.message(f"Available scripts from {source} (use --list for details):")
+            for name in scripts:
+                cc.message(f"  {name}")
+            return MessageResult("\nSpecify a script name to run, or use --list")
+        return MessageResult(
+            "No scripts defined. Add a 'scripts' section to snowflake.yml or manifest.yml"
+        )
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/commands.py]: You can first deal with simple case to reduce indentation levels
```suggestion
        if not scripts:
            return MessageResult(
                "No scripts defined. Add a 'scripts' section to snowflake.yml or manifest.yml"
            )
        source = manager.scripts_source or "project"
        cc.message(f"Available scripts from {source} (use --list for details):")
        for name in scripts:
            cc.message(f"  {name}")
         return MessageResult("\nSpecify a script name to run, or use --list")
```

File: src/snowflake/cli/_plugins/run/commands.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        script_name,
+        extra_args=extra_args,
+        var_overrides=vars_dict,
+        dry_run=dry_run,
+        verbose=ctx.verbose,
+        continue_on_error=continue_on_error,
+    )
+
+    if not result.success:
+        raise typer.Exit(result.exit_code)
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/commands.py]: I'm a bit biased here, as on one hand I'd prefer this minimalistic way of reporting failures, but on the other hand we're not telling what failed or why. Also it's not consistent with how we report errors

File: src/snowflake/cli/_plugins/run/manager.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+    def _parse_manifest_scripts(self, scripts_data: dict) -> Dict[str, ScriptModel]:
+        """Parse scripts section from manifest.yml into ScriptModel objects."""
+        result = {}
+        for name, script_def in scripts_data.items():
+            if isinstance(script_def, str):
+                result[name] = ScriptModel(cmd=script_def)
+            else:
+                result[name] = ScriptModel(**script_def)
+        return result
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/manager.py:127]: No error handling for malformed manifest.yaml. We've recently added `DCMManifest` class and afaik that's the only manifest we have right now, so might be better to use that or even extract some base fields such as manifest_version, project_type and scripts to a base class

File: src/snowflake/cli/_plugins/run/manager.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+        def replace_var(match: re.Match) -> str:
+            var_name = match.group(1)
+            if var_name in variables:
+                value = variables[var_name]
+                if shell_mode:
+                    return shlex.quote(value)
+                return value
+            log.warning("Variable '${%s}' not found, leaving as-is", var_name)
+            return match.group(0)
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/manager.py]: I'd rather raise an error here, as using the var name could lead to dangerous outcomes.

File: src/snowflake/cli/_plugins/run/manager.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+from snowflake.cli.api.console import cli_console as cc
+from snowflake.cli.api.constants import DEFAULT_SIZE_LIMIT_MB
+from snowflake.cli.api.project.schemas.project_definition import DefinitionV20
+from snowflake.cli.api.project.schemas.scripts import ScriptModel
+from snowflake.cli.api.secure_path import SecurePath
+from snowflake.cli.api.utils.models import ProjectEnvironment
+
+log = logging.getLogger(__name__)
+
+VARIABLE_PATTERN = re.compile(r"\$\{([^}]+)\}")
```
Comment (sfc-gh-jwilkowski) [inline src/snowflake/cli/_plugins/run/manager.py:39]: I'm worried about introducing another syntax for variable interpolation. Our users might get confused with those different syntax versions living side by side https://github.com/snowflakedb/snowflake-cli-templates/blob/main/streamlit_vnext_multi_page/snowflake.yml#L6-L9

**Review by sfc-gh-jwilkowski — COMMENTED** [review]
I'd like to request some changes. I think that some design decisions require more discussion with @sfc-gh-jbilek (like --list as flag rather than standalone command)

File: src/snowflake/cli/_plugins/run/manager.py (hunk 9 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/manager.py:127]: Good point on both fronts. The try/except already catches malformed YAML and logs it, so we won't crash on bad manifest files. Extracting shared base fields (manifest_version, project_type, scripts) from DCMManifest into a base class is a solid idea, but I think it's better suited for a follow-up PR to keep scope manageable here. Would that work?

File: src/snowflake/cli/_plugins/run/commands.py (hunk 1 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py:104]: Fair point on consistency. Scripts aren't Snowflake objects so `add_object_command_aliases` doesn't fit directly, but we could add a dedicated `snow run list` subcommand to follow the pattern. Worth discussing with @sfc-gh-jbilek as you suggested. For now, I've updated `--list` to return `CollectionResult` so `--format=json` works properly.

One callout... anything we put after "run" breaks a potential script. That is why I am hesitant on the subcommands

File: src/snowflake/cli/_plugins/run/manager.py (hunk 11 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/manager.py:39]: Valid concern. The distinction is: `<% ctx.env.XXX %>` is Jinja-style templating resolved at project load time (for entity definitions, identifiers, etc.), while `${var}` is resolved at script execution time and supports `--var` overrides for runtime flexibility. `${var}` is also the natural syntax users expect in shell commands (matches bash, Makefile, docker-compose conventions). I think clear documentation of the two scopes would address confusion. Happy to discuss if you'd prefer aligning to one syntax though.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 8 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Agreed, we should be more informative. Updated to raise `CliError` with the script name and exit code: `Script 'deploy' failed with exit code 1`. This is consistent with how other CLI commands report errors.

File: src/snowflake/cli/_plugins/run/manager.py (hunk 4 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/manager.py]: Good thought, but I'd prefer to keep single-source here. Having one file own all scripts avoids ambiguity about which file defines what, makes debugging simpler ("where is this script defined?" has exactly one answer), and prevents subtle ordering/shadowing issues if both files define the same name. The error message already guides users to pick one file.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 3 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Done, switched to `CliError` throughout.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 2 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Applied, thanks.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 5 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Went with `CollectionResult` to support `--format=json` and `--format=csv` as jwilkowski suggested.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 6 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Added `sanitize_for_terminal()` for script names displayed in the fallback path.

File: src/snowflake/cli/_plugins/run/commands.py (hunk 7 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/commands.py]: Applied the early-return pattern, thanks.

File: src/snowflake/cli/_plugins/run/manager.py (hunk 10 above)
Comment (sfc-gh-mborins) [reply src/snowflake/cli/_plugins/run/manager.py]: Agreed. Changed to raise `CliError` with a message pointing the user to define the variable or use `--var`.
